rust/lance/src/io/reader.rs

```diff
@@ -373,6 +375,42 @@ impl FileReader {
         self.metadata.is_empty()
     }
 
+    /// Stream desired full batches from the file.
+    ///
+    /// Parameters:
+    /// - **projection**: The schema of the returning [RecordBatch].
+    /// - **predicate**: A function that takes a batch ID and returns true if the batch should be
+    ///  returned.
+    ///
+    /// Returns:
+    /// - A stream of [RecordBatch]s.
+    #[allow(dead_code)]
+    pub(crate) fn batches_stream<'a>(
+        &'a self,
+        projection: Schema,
+        predicate: impl FnMut(&i32) -> bool + Send + Sync + 'a,
+    ) -> impl RecordBatchStream + 'a {
+        // Make projection an Arc so we can clone it and pass between threads.
+        let projection = Arc::new(projection);
+        let batches = (0..self.num_batches() as i32)
+            .filter(predicate)
+            .collect::<Vec<_>>();
```

Comment:
There any reason to collect the batches? I think we can just pass the iterator straight to `stream::iter()`.